SalesforceSDK/Analytics/Store/EventStoreManager.cs

```diff
@@ -0,0 +1,224 @@
+ * and/or other materials provided with the distribution.
+ * - Neither the name of salesforce.com, inc. nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission of salesforce.com, inc.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows;
+using Newtonsoft.Json.Linq;
+using Salesforce.SDK.Analytics.Model;
+using Salesforce.SDK.Core;
+using Salesforce.SDK.Logging;
+using PCLStorage;
+
+namespace Salesforce.SDK.Analytics.Store
+{
+    public class EventStoreManager : IEventStoreManager
+    {
+        private string _filenameSuffix;
+        private readonly IFolder _rootDir;
+        private string _encryptionKey;
+        private bool _isLoggingEnabled = true;
+        private int _maxEvents = 1000;
+        private static ILoggingService LoggingService => SDKServiceLocator.Get<ILoggingService>();
+        public string FilenameSuffix
+        {
+            get { return _filenameSuffix; }
+            set { _filenameSuffix = value; }
+        } 
+
+        public string EncryptionKey
+        {
+            get { return _encryptionKey; }
+            set { _encryptionKey = value; }
+        }
+
+        public EventStoreManager(string fileNameSuffix, string encryptionKey)
+        {
+            FilenameSuffix = fileNameSuffix;
+            EncryptionKey = encryptionKey;
+            _rootDir = FileSystem.Current.LocalStorage;
+        }
+
+        public async Task StoreEventAsync(InstrumentationEvent instrumentationEvent)
+        {
+            if (instrumentationEvent == null || string.IsNullOrEmpty(instrumentationEvent.ToJson().ToString()))
+            {
+                LoggingService.Log("Invalid Event", LoggingLevel.Error);
+                return;
+            }
+            //TODO: Add check for shouldstoreevent
+            var fileName = instrumentationEvent.EventId + _filenameSuffix;
+            //Open file
+            IFile file = await _rootDir.CreateFileAsync(fileName, CreationCollisionOption.OpenIfExists);
```

Comment:
tip: you can increase perf for consumers if your first `await` call in any async method is followed with ConfigureAwait(false).  That basically tells the thread synchronizer that when it returns from running this async method it can return on a non-UI thread.  This is good in case some viewmodel on the UI thread called into your service here because the rest of this method doesn't need to be on the UI thread.  It's a bit nuanced and feel free to come talk to me for more details, but basically if you have an async method that doesn't need to run on the UI thread (which is almost anything that isn't a View or ViewModel) then it's a good idea to call ConfigureAwait(false)... so this line would look like `IFile file = await _rootDir.CreateFileAsync(fileName, CreationCollisionOption.OpenIfExists).ConfigureAwait(false);`

You really only need to do it on the first awaited call in your method as that will take it off the UI thread (if it's on the UI thread)... but it doesn't to do that on every await.

[Stephen Cleary](http://blog.stephencleary.com/2012/02/async-and-await.html) is the master of async .net code, he has a [good writeup on ConfigureAwait](http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html) if you want more info.